### PR TITLE
Add globlal WAR/JAR deploy dialog (classes only, nothing visible or wired)

### PR DESCRIPTION
**Pull request by chanseokoh**
Adds two small classes:
- `FlexExistingArtifactDeployPreferencesDialog`
- `FlexExistingArtifactDeployPreferencesPanel`

The dialog will add one additional WAR/JAR file input field to the current flex deploy dialog. This field is placed immediately above the app.yaml input field. (Again, this PR just adds the dialog. It is not visible.)

selection_015

Other things are mostly opportunistic refactoring.

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesDialog.java [hunk 1]
```diff
@@ -66,8 +66,8 @@ protected Control createDialogArea(final Composite parent) {
   @Override
   protected AppEngineDeployPreferencesPanel createDeployPreferencesPanel(Composite container,
       IProject project, IGoogleLoginService loginService, Runnable layoutChangedHandler,
-      boolean requireValues, ProjectRepository projectRepository) {
+      ProjectRepository projectRepository) {
     return new FlexDeployPreferencesPanel(container, project, loginService, layoutChangedHandler,
-        requireValues, projectRepository);
+        true /* requireValues */, projectRepository);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesDialog.java]: or just introduce a `requireValues=true` local variable

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java [hunk 2]
```diff
@@ -38,20 +39,25 @@
...
+
   @Override
   protected void createCenterArea() {
-    IPath fileFieldBasePath = project.getLocation();
-
     Text appYamlField = createBrowseFileRow(
         Messages.getString("deploy.preferences.dialog.label.app.yaml"),
         Messages.getString("tooltip.app.yaml"),
-        fileFieldBasePath, new String[] {"*.yaml"});
+        getFileFieldBasePath(), new String[] {"*.yaml"});
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java]: This method name ` getFileFieldBasePath` isn't really obvious to me. Any other ideas?

Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java]: I can add a Javadoc.

Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java]: I'd rather rename so it's more obvious. I completely misunderstood what this did on first read.

Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java]: `getDialogBasePathForFileFields`?

Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java]: That's going in the wrong direction. :-)  Try not to reference the UI here. E.g. `getWorkingDirectory()`

Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java]: It's not a working directory for something, but a path that will be automatically added as a prefix if users enter a non-absolute path in file input fields in the dialog. For example, if a user enters `src/main/appengine/app.yaml` and this base path is an absolute path of a project `/home/user/workspace/my-project`, we can construct `/home/user/workspace/my-project/src/main/appengine/app.yaml`.

Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployPreferencesPanel.java]: On second thought, `getWorkingDirectory()` is good enough.

**Comment by chanseokoh** [comment]
I signed it!

**Pull request merged**